Add biased-decoding feature

scotfang opened the pull request.

scotfang commented:
Hi, this PR contains all the code and tests needed for the new anti-flicker feature I've implemented.  Please take a look when you have a chance, no rush!

guillaumekln commented on src/decoding.cc:
```diff
@@ -113,11 +115,78 @@ namespace ctranslate2 {
+
+    // Scalar's need to be allocated on CPUs.
+    StorageView scalar_discount(1 - prefix_bias_beta, Device::CPU);
+    // const dim_t cur_beam_size = is_expanded ? _beam_size : 1;
```
Remove this line?

scotfang replied:
done

guillaumekln commented on src/decoding.cc:
```diff
@@ -113,11 +115,78 @@ namespace ctranslate2 {
+  }
 
-  BeamSearch::BeamSearch(const dim_t beam_size, const float length_penalty, const float coverage_penalty)
+  BeamSearch::BeamSearch(const dim_t beam_size, const float length_penalty, const float coverage_penalty, 
+                         const float prefix_bias_beta)
```
Minor formatting suggestion:

```suggestion
  BeamSearch::BeamSearch(const dim_t beam_size,
                         const float length_penalty,
                         const float coverage_penalty, 
                         const float prefix_bias_beta)
```

guillaumekln commented on src/decoding.cc:
```diff
@@ -189,18 +267,34 @@ namespace ctranslate2 {
+
+      if (bias_towards_prefix) {
+        if (!biased_decoder) {
+          biased_decoder.reset(new BiasedDecoder());
```
We try to follow this pattern for unique_ptr:

```suggestion
          biased_decoder = std::make_unique<BiasedDecoder>();
```

guillaumekln commented on tests/translator_test.cc:
```diff
@@ -338,6 +339,295 @@ TEST(TranslatorTest, TranslateBatchWithPrefixAndEmpty) {
   EXPECT_EQ(result[4].output(), (std::vector<std::string>{"a", "z", "z", "a"}));
 }
 
+TEST(TranslatorTest, TranslateBatchWithStronglyBiasedPrefixAndEmpty) {
```
Empty inputs are handled outside of the decoding code, so I don't think it is relevant to include empty inputs when testing this feature.

I suggest removing the empty input and renaming the test case.

scotfang replied:
done

guillaumekln commented on tests/translator_test.cc:
```diff
@@ -338,6 +339,295 @@ TEST(TranslatorTest, TranslateBatchWithPrefixAndEmpty) {
+  EXPECT_EQ(result[4].output(), (std::vector<std::string>{"a", "z", "z", "a"}));
+}
+
+TEST(TranslatorTest, TranslateBatchWithWeaklyBiasedPrefixAndEmpty) {
```
Same comment regarding the empty input.

scotfang replied:
done

guillaumekln commented on src/decoding.cc:
```diff
@@ -113,11 +115,78 @@ namespace ctranslate2 {
+    std::vector<StorageView*> logit_beam_views(num_beams);
+    std::vector<StorageView> log_prob_beam_view_storage(num_beams, StorageView(device, dtype));
+    std::vector<StorageView*> log_prob_beam_views(num_beams);
+    for (int i = 0; i < num_beams; ++i) {
```
```suggestion
    for (dim_t i = 0; i < num_beams; ++i) {
```

guillaumekln commented on src/decoding.cc:
```diff
@@ -170,6 +239,15 @@ namespace ctranslate2 {
       attention->resize(batch_size);
     }
 
+    // TODO(scotfang): Instead of initializing and empty vector, use unique_ptr.
```
I don't think initializing an empty vector will cause any performance issue, so you can probably remove this comment.

scotfang replied:
done

guillaumekln commented:
Can you also update the relevant documentation?

* Add the feature in README.md -> Supported decoding options
* Add the new arguments in the Python API reference in docs/python.md
* (optional) Add an example at the end of docs/decoding.md

scotfang commented on include/ctranslate2/translator.h:
```diff
@@ -30,6 +30,15 @@ namespace ctranslate2 {
     float length_penalty = 0;
     // Coverage value to apply during beam search.
     float coverage_penalty = 0;
+    // Biases decoding towards a given prefix, see https://arxiv.org/abs/1912.03393 --section 4.2
+    // Only activates biased-decoding when beta is in range (0, 1) and SearchStrategy is set to BeamSearch.
+    //
+    // Setting beta to 1.0 is equivalent to using a hard prefix, setting it to 0 is equivalent
+    // to unconstrained decoding.  The higher beta is, the stronger the bias towards the given prefix.
+    //
+    // If you translate with a non-empty prefix and beta <= 0, 
+    // then the prefix will be used as a hard-prefix rather than a soft, biased-prefix.
```
Made docs more clear.

scotfang commented:
Addressed issues brought up in review, re-requested review.  

Should I also try to resolve conflicts? Some of the changes from master look new to me.

guillaumekln reviewed (approved):
Looks good! I resolved the conflicts in `decoding.cc` and apply some minor reformatting to match the existing code.

Thanks for your contribution!

scotfang commented:
Thank you so much for merging this work!
